cmd/ipfs/main.go

```diff
@@ -297,70 +296,106 @@ func callPreCommandHooks(ctx context.Context, details cmdDetails, req cmds.Reque
 	return nil
 }
 
-func callCommand(ctx context.Context, req cmds.Request, root *cmds.Command, cmd *cmds.Command) (cmds.Response, error) {
+func callCommand(ctx context.Context, req cmds.Request, root *cmds.Command, cmd *cmds.Command) error {
 	log.Info(config.EnvDir, " ", req.InvocContext().ConfigRoot)
-	var res cmds.Response
 
 	err := req.SetRootContext(ctx)
 	if err != nil {
-		return nil, err
+		return err
 	}
 
 	details, err := commandDetails(req.Path(), root)
 	if err != nil {
-		return nil, err
+		return err
 	}
 
 	client, err := commandShouldRunOnDaemon(*details, req, root)
 	if err != nil {
-		return nil, err
+		return err
 	}
 
 	err = callPreCommandHooks(ctx, *details, req, root)
 	if err != nil {
-		return nil, err
+		return err
+	}
+
+	encTypeStr, found, err := req.Option("encoding").String()
+	if !found || err != nil {
+		log.Error("error getting encoding - using JSON. reason: ", err)
+		encTypeStr = "json"
+	}
+	encType := cmds.EncodingType(encTypeStr)
+
+	var (
+		re     cmds.ResponseEmitter
+		exitCh <-chan int
+	)
+
+	// first if condition checks the command's encoder map, second checks global encoder map (cmd vs. cmds)
+	if enc, ok := cmd.Encoders[encType]; ok {
+		re, exitCh = cli.NewResponseEmitter(os.Stdout, os.Stderr, enc, req)
+	} else if enc, ok := cmds.Encoders[encType]; ok {
+		re, exitCh = cli.NewResponseEmitter(os.Stdout, os.Stderr, enc, req)
+	} else {
+		return fmt.Errorf("could not find matching encoder for enctype %#v", encType)
 	}
 
 	if cmd.PreRun != nil {
 		err = cmd.PreRun(req)
 		if err != nil {
-			return nil, err
+			return err
 		}
 	}
 
+	if cmd.PostRun != nil && cmd.PostRun[cmds.CLI] != nil {
+		re = cmd.PostRun[cmds.CLI](req, re)
+	}
+
 	if client != nil && !cmd.External {
 		log.Debug("executing command via API")
-		res, err = client.Send(req)
+
+		res, err := client.Send(req)
 		if err != nil {
 			if isConnRefused(err) {
 				err = repo.ErrApiNotRunning
 			}
-			return nil, wrapContextCanceled(err)
+
+			return wrapContextCanceled(err)
 		}
 
+		go func() {
+			err = cmds.Copy(re, res)
+			if err != nil {
+				re.SetError(err, cmdkit.ErrNormal|cmdkit.ErrFatal)
+			}
+		}()
 	} else {
 		log.Debug("executing command locally")
 
 		pluginpath := filepath.Join(req.InvocContext().ConfigRoot, "plugins")
 		if _, err := loader.LoadPlugins(pluginpath); err != nil {
-			return nil, err
+			return err
 		}
 
 		err := req.SetRootContext(ctx)
 		if err != nil {
-			return nil, err
+			return err
 		}
 
 		// Okay!!!!! NOW we can call the command.
-		res = root.Call(req)
-
+		go func() {
+			err := root.Call(req, re)
```

Comment:
why is this in a goroutine now? We end up waiting for things to complete below anyways.

Reply:
Because otherwise `cli.ResponeEmitter` would try to write to `exitCh` (which is returned by `cli.NewResponseEmitter`, main.go:336), which is only read a few lines later. Should I use a buffered channel here?